### PR TITLE
[Merged by Bors] - refactor: make `(Left|Right)Commutative` typeclasses

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
`LeftCommutative` and `RightCommutative` are currently pure definitions, but they are used in other declarations as if they were typeclasses. As they are not actually typeclasses they have to be passed in explicitly, which causes unnecessary repetition and code bloat.

This PR makes them typeclasses and adds supporting instances: two converted from the theorems `left_comm` and `right_comm` and two others extracted from `Data.Multiset.Basic`. In #16748 they will be moved with the very structurally similar typeclass `IsSymmOp` out of `Init.Logic` to a new file.

The shaking of this PR revealed some uses of lemmas deprecated in #16438 but not caught by the linter. We fix these too.

### CODE REVIEW

File: Mathlib/Init/Algebra/Classes.lean [hunk 1]
```diff
@@ -180,7 +180,7 @@ theorem lt_of_lt_of_incomp {α : Sort u} {lt : α → α → Prop} [IsStrictWeak
     [DecidableRel lt] : ∀ {a b c}, lt a b → ¬lt b c ∧ ¬lt c b → lt a c :=
   @fun a b c hab ⟨nbc, ncb⟩ =>
   have nca : ¬lt c a := fun hca => absurd (trans_of lt hca hab) ncb
-  Decidable.by_contradiction fun nac : ¬lt a c =>
+  Decidable.byContradiction fun nac : ¬lt a c =>
```
Comment (YaelDillies) [inline Mathlib/Init/Algebra/Classes.lean:183]: Does this have anything to do with the PR?

**Review by YaelDillies — COMMENTED** [review]
This looks reasonable, but can you explain the motivation?

File: Mathlib/Init/Algebra/Classes.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Init/Algebra/Classes.lean:183]: It is the fallout of the shake linter.

**Comment by YaelDillies** [comment]
So... your motivation for making them a typeclass is to move them to a new file about typeclasses? That doesn't seem very convincing.

**Comment by Parcly-Taxel** [comment]
> So... your motivation for making them a typeclass is to move them to a new file about typeclasses? That doesn't seem very convincing.

Structurally `Left/RightCommutative` are very similar to `IsSymmOp`. Furthermore making them typeclasses allows – as can be seen in the PR – to golf some code.

**Review by YaelDillies — APPROVED** [review]
That is much better motivation!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**